L1Trigger/TrackFindingTracklet/plugins/ProducerKFin.cc

```diff
@@ -0,0 +1,261 @@
+    const DataFormat& dfQoverPt = dataFormats_->format(Variable::qOverPt, Process::sf);
+    const DataFormat& dfPhiT = dataFormats_->format(Variable::phiT, Process::sf);
+    const DataFormat& dfPhi = dataFormats_->format(Variable::phi, Process::sf);
+    const DataFormat& dfZ = dataFormats_->format(Variable::z, Process::sf);
+    const int numStreamsTracks = setup_->numRegions() * trackBuilderChannel_->numChannels();
+    const int numStreamsStubs = numStreamsTracks * setup_->numLayers();
+    // empty KFin products
+    TTDTC::Streams streamAcceptedStubs(numStreamsStubs);
+    StreamsTrack streamAcceptedTracks(numStreamsTracks);
+    TTDTC::Streams streamLostStubs(numStreamsStubs);
+    StreamsTrack streamLostTracks(numStreamsTracks);
+    // read in hybrid track finding product and produce KFin product
+    if (setup_->configurationSupported()) {
+      Handle<TTTracks> handleTTTracks;
+      iEvent.getByToken<TTTracks>(edGetTokenTTTracks_, handleTTTracks);
+      const TTTracks& ttTracks = *handleTTTracks.product();
+      vector<TTTrackRefs> ttTrackRefsStreams(numStreamsTracks);
+      vector<int> nTTTracksStreams(numStreamsTracks, 0);
+      int channelId;
+      for (const TTTrack<Ref_Phase2TrackerDigi_>& ttTrack : ttTracks)
+        if (trackBuilderChannel_->channelId(ttTrack, channelId))
+          nTTTracksStreams[channelId]++;
+      channelId = 0;
+      for (int nTTTracksStream : nTTTracksStreams)
+        ttTrackRefsStreams[channelId++].reserve(nTTTracksStream);
+      int i(0);
+      for (const TTTrack<Ref_Phase2TrackerDigi_>& ttTrack : ttTracks)
+        if (trackBuilderChannel_->channelId(ttTrack, channelId))
+          ttTrackRefsStreams[channelId].emplace_back(TTTrackRef(handleTTTracks, i++));
+      for (channelId = 0; channelId < numStreamsTracks; channelId++) {
+        const TTTrackRefs& ttTrackRefs = ttTrackRefsStreams[channelId];
+        vector<TrackKFin> tracks;
+        tracks.reserve(ttTrackRefs.size());
+        vector<StubKFin> stubs;
+        const int nStubs = accumulate(ttTrackRefs.begin(), ttTrackRefs.end(), 0, [](int& sum, const TTTrackRef& ttTrackRef){ return sum += ttTrackRef->getStubRefs().size(); });
+        stubs.reserve(nStubs);
+        int trackId(0);
+        vector<int> numLayerStubs(setup_->numLayers(), 0);
+        for (const TTTrackRef& ttTrackRef : ttTrackRefs) {
+          // cut on more then 256 tracks
+          if (trackId >= dataFormats_->format(Variable::trackId, Process::kfin).range())
+            continue;
+          // get rz parameter
+          double cot = ttTrackRef->tanL();
+          double zT = ttTrackRef->z0() + setup_->chosenRofZ() * cot;
+          int binEta(-1);
+          for (; binEta < setup_->numSectorsEta(); binEta++)
+            if (zT < sinh(setup_->boundarieEta(binEta + 1)) * setup_->chosenRofZ())
+              break;
+          // cut on outer eta sector boundaries
+          if (binEta == -1 || binEta == setup_->numSectorsEta())
+            continue;
```

Comment:
The tracklet algo finds a significant number of tracks beyond abs(eta) = 2.4, (since it only requires 4 tracker layers). Will this cut kill all those outside this range?

Reply:
yes

Reply:
Understood. But we should check if this cut causes any efficiency loss at high eta.

Reply:
The current total loss where this cut will contribute to was measured to 1.5 permille from chris. that sounded acceptable to me.